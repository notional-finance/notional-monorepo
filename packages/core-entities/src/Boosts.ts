import {
  Network,
  SECONDS_IN_DAY,
  floorToMidnight,
  getNowSeconds,
} from '@notional-finance/util';
import { TokenDefinition } from './Definitions';
import { TokenBalance } from './token-balance';
import { Registry } from './Registry';

export function getPointsPerDay(t: TokenBalance) {
  return (
    t.toUnderlying().abs().toFiat('USD').toFloat() *
    getArbBoosts(t.token, t.isNegative())
  );
}

export function getArbBoosts(b: TokenDefinition, isDebt: boolean) {
  if (b.network !== Network.arbitrum) return 0;

  if (b.tokenType === 'VaultShare') {
    // ezETH/wstETH
    if (b.vaultAddress === '0xd7c3dc1c36d19cf4e8cea4ea143a2f4458dd1937')
      return 0;
    return 8;
  } else if (
    b.tokenType === 'PrimeDebt' ||
    (b.tokenType === 'PrimeCash' && isDebt)
  ) {
    // Variable Borrow
    if (b.currencyId === 2) return 4; // DAI
    if (b.currencyId === 3) return 4; // USDC
    if (b.currencyId === 6) return 4; // FRAX
    if (b.currencyId === 8) return 4; // USDT
  } else if (b.tokenType === 'fCash' && isDebt) {
    // Fixed Borrow
    if (b.currencyId === 2) return 4; // DAI
    if (b.currencyId === 3) return 4; // USDC
    if (b.currencyId === 6) return 4; // FRAX
    if (b.currencyId === 8) return 4; // USDT
  } else if (b.tokenType === 'PrimeCash' && !isDebt) {
    // Variable Lend
    if (b.currencyId === 1) return 2; // ETH
    if (b.currencyId === 3) return 2; // USDC

    if (b.currencyId === 8) return 1; // USDT
    if (b.currencyId === 2) return 1; // DAI
    if (b.currencyId === 6) return 1; // FRAX
    if (b.currencyId === 5) return 1; // wstETH
  } else if (b.tokenType === 'fCash' && !isDebt) {
    // Fixed Lend
    if (b.currencyId === 1) return 2; // ETH
    if (b.currencyId === 3) return 2; // USDC

    if (b.currencyId === 8) return 1; // USDT
    if (b.currencyId === 2) return 1; // DAI
    if (b.currencyId === 6) return 1; // FRAX
    if (b.currencyId === 5) return 1; // wstETH
  }

  return 0;
}

export function getPointsAPY(
  boost: number,
  currentTotalPoints: number,
  totalARB: number,
  seasonStart: Date,
  seasonEnd: Date
) {
  const totalARBPerSeason = TokenBalance.fromFloat(
    totalARB,
    Registry.getTokenRegistry().getTokenBySymbol(Network.arbitrum, 'ARB')
  );

  const daysIntoSeason = Math.floor(
    (floorToMidnight(getNowSeconds()) - seasonStart.getTime() / 1000) /
      SECONDS_IN_DAY
  );
  const daysLeftInSeason = Math.ceil(
    Math.max(seasonEnd.getTime() / 1000 - floorToMidnight(getNowSeconds()), 0) /
      SECONDS_IN_DAY
  );
  const dailyEmissionRate = currentTotalPoints / daysIntoSeason;
  const projectedPointTotal =
    currentTotalPoints + dailyEmissionRate * daysLeftInSeason;
  const pointsPerDollar = boost;
  const arbSharePerDollar = pointsPerDollar / projectedPointTotal;

  const dailyArbPerDollarInUSD =
    totalARBPerSeason.toFiat('USD').toFloat() * arbSharePerDollar;

  const result = dailyArbPerDollarInUSD * 365 * 100

  // TODO: remove this when we get the season two points calculating correctly
  return result > 0 ? result : 0;
}
