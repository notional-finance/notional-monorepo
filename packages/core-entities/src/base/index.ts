export { BaseRegistry } from './base-registry';
