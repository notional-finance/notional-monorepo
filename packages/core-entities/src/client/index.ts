export { TokenRegistryClient } from './token-registry-client';
export { ConfigurationClient } from './configuration-client';
export { NOTERegistryClient } from './note-registry-client';
