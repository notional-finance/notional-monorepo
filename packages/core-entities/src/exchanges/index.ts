import MetaStablePool from './BalancerV2/meta-stable-pool';
import WeightedPool from './BalancerV2/weighted-pool';
import BaseLiquidityPool from './base-liquidity-pool';
import Curve2TokenPoolV1 from './Curve/Curve2TokenPoolV1';
import Curve2TokenPoolV2 from './Curve/Curve2TokenPoolV2';
import {
  TwoTokenComposableStablePool,
  ThreeTokenComposableStablePool,
} from './BalancerV2/composable-stable-pool';
import { TokenBalance } from '../token-balance';
import { Network } from '@notional-finance/util';
import { fCashMarket } from './NotionalV3/fCash-market';

type PoolConstructor = new (
  network: Network,
  balances: TokenBalance[],
  totalSupply: TokenBalance,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  poolParams: any
) => BaseLiquidityPool<unknown>;

const PoolClasses: Record<
  string,
  typeof BaseLiquidityPool<unknown> & PoolConstructor
> = {
  MetaStablePool: MetaStablePool,
  WeightedPool: WeightedPool,
  fCashMarket: fCashMarket,
  TwoTokenComposableStablePool: TwoTokenComposableStablePool,
  ThreeTokenComposableStablePool: ThreeTokenComposableStablePool,
};

export {
  MetaStablePool,
  WeightedPool,
  BaseLiquidityPool,
  TwoTokenComposableStablePool,
  ThreeTokenComposableStablePool,
  PoolClasses,
  PoolConstructor,
  Curve2TokenPoolV1,
  Curve2TokenPoolV2,
};
