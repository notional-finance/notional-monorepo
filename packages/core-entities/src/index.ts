export * from './token-balance';
export * from './Definitions';
export * from './Matchers';
export * from './Registry';
export { AccountFetchMode } from './client/account-registry-client';
export * from './exchanges';
export * from './vaults';
export { ConfigurationClient, NOTERegistryClient } from './client';
export * from './config/fiat-config';
export * from './server';
export { ClientRegistry } from './client/client-registry';
export type { ServerRegistry } from './server/server-registry';
export type { GraphDocument } from './server/analytics-server';
export {
  fetchGraph,
  loadGraphClientDeferred,
  fetchGraphPaginate,
} from './server/server-registry';
export * from './config/whitelisted-vaults';
export { DeprecatedVaults } from './server/vault-overrides';
export { SecondaryIncentiveToken } from './config/whitelisted-tokens';
export * from './Boosts';

export { NetworkClientModel, NetworkServerModel } from './models/NetworkModel';
