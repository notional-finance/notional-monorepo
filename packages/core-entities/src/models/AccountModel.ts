import { flow, types } from 'mobx-state-tree';
import { NotionalTypes } from './ModelTypes';
import { getProviderFromNetwork } from '@notional-finance/util';
import { providers } from 'ethers';
import { fetchCurrentAccount } from '../client/accounts/current-account';
import { AccountDefinition, CacheSchema } from '../Definitions';

const AccountIncentiveDebtModel = types.model('AccountIncentiveDebt', {
  value: NotionalTypes.TokenBalance,
  currencyId: types.number,
});

const BalanceStatementModel = types.model('BalanceStatement', {
  token: NotionalTypes.TokenDefinition,
  blockNumber: types.number,
  underlying: types.late(() => NotionalTypes.TokenDefinition),
  currentBalance: NotionalTypes.TokenBalance,
  adjustedCostBasis: NotionalTypes.TokenBalance,
  totalILAndFees: NotionalTypes.TokenBalance,
  totalProfitAndLoss: NotionalTypes.TokenBalance,
  totalInterestAccrual: NotionalTypes.TokenBalance,
  accumulatedCostRealized: NotionalTypes.TokenBalance,
  incentives: types.array(
    types.model({
      totalClaimed: NotionalTypes.TokenBalance,
      adjustedClaimed: NotionalTypes.TokenBalance,
    })
  ),
  impliedFixedRate: types.maybe(types.number),
});

const AccountHistoryModel = types.model('AccountHistory', {
  label: types.string,
  txnLabel: types.maybe(types.string),
  timestamp: types.number,
  blockNumber: types.number,
  token: NotionalTypes.TokenDefinition,
  underlying: NotionalTypes.TokenDefinition,
  tokenAmount: NotionalTypes.TokenBalance,
  bundleName: types.string,
  transactionHash: types.string,
  underlyingAmountRealized: NotionalTypes.TokenBalance,
  underlyingAmountSpot: NotionalTypes.TokenBalance,
  realizedPrice: NotionalTypes.TokenBalance,
  spotPrice: NotionalTypes.TokenBalance,
  vaultName: types.maybe(types.string),
  impliedFixedRate: types.maybe(types.number),
  isTransientLineItem: types.boolean,
  account: types.maybe(types.string),
});

export const AccountModel = types
  .model('Account', {
    address: types.string,
    network: NotionalTypes.Network,
    lastUpdateTimestamp: types.optional(types.number, 0),
    systemAccountType: types.optional(NotionalTypes.SystemAccount, 'None'),
    balances: types.optional(types.array(NotionalTypes.TokenBalance), []),
    allowPrimeBorrow: types.maybe(types.boolean),
    vaultLastUpdateTime: types.optional(types.map(types.number), {}),
    accountIncentiveDebt: types.optional(
      types.array(AccountIncentiveDebtModel),
      []
    ),
    secondaryIncentiveDebt: types.optional(
      types.array(AccountIncentiveDebtModel),
      []
    ),
    allowances: types.optional(
      types.array(
        types.model({
          spender: types.string,
          amount: NotionalTypes.TokenBalance,
        })
      ),
      []
    ),
    stakeNOTEStatus: types.maybe(
      types.model({
        inCoolDown: types.boolean,
        inRedeemWindow: types.boolean,
        redeemWindowBegin: types.number,
        redeemWindowEnd: types.number,
      })
    ),
    accountHistory: types.maybe(types.array(AccountHistoryModel)),
    balanceStatement: types.maybe(types.array(BalanceStatementModel)),
    historicalBalances: types.maybe(
      types.array(
        types.model({
          timestamp: types.number,
          balance: NotionalTypes.TokenBalance,
        })
      )
    ),
  })
  .actions((self) => {
    let provider = getProviderFromNetwork(self.network);

    const refreshAccount = flow(function* () {
      const result: CacheSchema<AccountDefinition> = yield fetchCurrentAccount(
        self.network,
        self.address,
        provider
      );

      const accountDefinition =
        result.values.length === 1 &&
        result.values[0].length === 2 &&
        result.values[0][1] !== null
          ? result.values[0][1]
          : undefined;

      if (accountDefinition) {
        // This is supposed to do a diff and only update the ones that have changed
        self.balances.replace(accountDefinition.balances);

        if (accountDefinition.allowances) {
          self.allowances.replace(accountDefinition.allowances);
        }
        if (accountDefinition.systemAccountType) {
          self.systemAccountType = accountDefinition.systemAccountType;
        }
        if (accountDefinition.allowPrimeBorrow) {
          self.allowPrimeBorrow = accountDefinition.allowPrimeBorrow;
        }
        if (accountDefinition.vaultLastUpdateTime) {
          self.vaultLastUpdateTime.replace(
            accountDefinition.vaultLastUpdateTime
          );
        }
        if (accountDefinition.accountIncentiveDebt) {
          self.accountIncentiveDebt.replace(
            accountDefinition.accountIncentiveDebt
          );
        }
        if (accountDefinition.secondaryIncentiveDebt) {
          self.secondaryIncentiveDebt.replace(
            accountDefinition.secondaryIncentiveDebt
          );
        }
        if (accountDefinition.stakeNOTEStatus) {
          self.stakeNOTEStatus = accountDefinition.stakeNOTEStatus;
        }
        self.lastUpdateTimestamp = result.lastUpdateTimestamp;
      }
    });

    return {
      afterCreate: refreshAccount,
      setProvider: (p: providers.Provider) => {
        provider = p;
      },
    };
  });
