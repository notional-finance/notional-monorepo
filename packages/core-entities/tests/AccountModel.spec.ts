import { Registry, AccountFetchMode } from '../src';
import { AccountModel } from '../src/models/AccountModel';
import { Network } from '@notional-finance/util';

describe('AccountModel', () => {
  beforeAll(async () => {
    Registry.initialize(
      {
        NX_SUBGRAPH_API_KEY: 'test',
      },
      'https://registry.notional.finance',
      AccountFetchMode.SINGLE_ACCOUNT_DIRECT,
      false,
      false,
      true
    );

    await Registry.triggerRefresh(Network.mainnet);
  });

  it('creates an account model with required fields and matches snapshot', () => {
    const account = AccountModel.create({
      address: '0xd74e7325dFab7D7D1ecbf22e6E6874061C50f243',
      network: Network.mainnet,
    });

    expect(account).toMatchSnapshot();
  });
});
