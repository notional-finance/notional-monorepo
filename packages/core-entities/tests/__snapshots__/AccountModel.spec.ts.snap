// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountModel creates an account model with required fields and matches snapshot 1`] = `
Object {
  "accountHistory": undefined,
  "accountIncentiveDebt": Array [
    Object {
      "currencyId": 1,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
    Object {
      "currencyId": 0,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
    },
  ],
  "address": "0xd74e7325dFab7D7D1ecbf22e6E6874061C50f243",
  "allowPrimeBorrow": true,
  "allowances": Array [
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
        "network": "mainnet",
        "tokenId": "0x0000000000000000000000000000000000000000",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
        "network": "mainnet",
        "tokenId": "0x40d16fc0246ad3160ccc09b8d0d3a2cd28ae6c2f",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x6b175474e89094c44da98b954eedeac495271d0f",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x83f20f44975d03b1b09e64809b757c47f942beea",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x853d955acef822db058eb8505911ed77f175b99e",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffc6c78ff",
        "network": "mainnet",
        "tokenId": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0xfffffffffffffffffffffffffffffffffffffffffffffffffe897fb1b585a3fe",
        "network": "mainnet",
        "tokenId": "0xae78736cd615f374d3085123a210448e74fc6393",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0xbe9895146f7af43049ca1c1ae358b0541ea49704",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0xffffffffffffffe8b78917ff",
        "network": "mainnet",
        "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
        "network": "mainnet",
        "tokenId": "0xdac17f958d2ee523a2206206994597c13d831ec7",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
    Object {
      "amount": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x38de42f4ba8a35056b33a746a6b45be9b1c3b9d2",
      },
      "spender": "0x6e7058c91F85E0F6db4fc9da2CA41241f5e4263f",
    },
  ],
  "balanceStatement": undefined,
  "balances": Array [
    Object {
      "_isTokenBalance": true,
      "hex": "0x11d05f",
      "network": "mainnet",
      "tokenId": "0x4a9e282635567cc4d3c6a24e16c2335f10dee9b8",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x9661ea1fa5176e",
      "network": "mainnet",
      "tokenId": "0x0000000000000000000000000000000000000000",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x8b43078288cc4ce6",
      "network": "mainnet",
      "tokenId": "0x40d16fc0246ad3160ccc09b8d0d3a2cd28ae6c2f",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x02fe5486ee56942ab6",
      "network": "mainnet",
      "tokenId": "0x6b175474e89094c44da98b954eedeac495271d0f",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x7f39c581f595b53c5cb19bd0b3f8da6c935e2ca0",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x83f20f44975d03b1b09e64809b757c47f942beea",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x853d955acef822db058eb8505911ed77f175b99e",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x28e360",
      "network": "mainnet",
      "tokenId": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0xae78736cd615f374d3085123a210448e74fc6393",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0xbe9895146f7af43049ca1c1ae358b0541ea49704",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x0252f499e0",
      "network": "mainnet",
      "tokenId": "0xcfeaead4947f0705a14ec42ac3d44129e1ef3ed5",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x08a8bd3e",
      "network": "mainnet",
      "tokenId": "0xdac17f958d2ee523a2206206994597c13d831ec7",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x034e68198bd4f48d3a",
      "network": "mainnet",
      "tokenId": "0x38de42f4ba8a35056b33a746a6b45be9b1c3b9d2",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x26aed4c3",
      "network": "mainnet",
      "tokenId": "0x0000000030fba4a7ec8591f25b4d37fd79943a4bb6e553e2000bffffffffff09",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x0000000030fba4a7ec8591f25b4d37fd79943a4bb6e553e2000bffffffffff0a",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x45ed6f",
      "network": "mainnet",
      "tokenId": "0x0000000032d82a1c8618c7be7fe85b2f1c44357a871d52d10007ffffffffff09",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x0000000032d82a1c8618c7be7fe85b2f1c44357a871d52d10007ffffffffff0a",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x5ae97053",
      "network": "mainnet",
      "tokenId": "0x0000000084e58d8faa4e3b74d55d9fc762230f15d95570b80003ffffffffff09",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "-0x4c6b5354",
      "network": "mainnet",
      "tokenId": "0x0000000084e58d8faa4e3b74d55d9fc762230f15d95570b80003ffffffffff0a",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x267f2333",
      "network": "mainnet",
      "tokenId": "0x00000000b1113cf888a019693b254da3d90f841072d85172000bffffffffff09",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x00000000b1113cf888a019693b254da3d90f841072d85172000bffffffffff0a",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x023b529030",
      "network": "mainnet",
      "tokenId": "0x00000000d6aa58cf21a0edb33375d6c0434b8bb5b589f0210003ffffffffff09",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "-0x01a6c720c0",
      "network": "mainnet",
      "tokenId": "0x00000000d6aa58cf21a0edb33375d6c0434b8bb5b589f0210003ffffffffff0a",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x52d528",
      "network": "mainnet",
      "tokenId": "0x00000000f94507f3dece4cc4c73b6cf228912b85eadc9cfb0001ffffffffff09",
    },
    Object {
      "_isTokenBalance": true,
      "hex": "0x00",
      "network": "mainnet",
      "tokenId": "0x00000000f94507f3dece4cc4c73b6cf228912b85eadc9cfb0001ffffffffff0a",
    },
  ],
  "historicalBalances": undefined,
  "lastUpdateTimestamp": 1723847447,
  "network": "mainnet",
  "secondaryIncentiveDebt": Array [
    Object {
      "currencyId": 11,
      "value": Object {
        "_isTokenBalance": true,
        "hex": "0x00",
        "network": "mainnet",
        "tokenId": "0x40d16fc0246ad3160ccc09b8d0d3a2cd28ae6c2f",
      },
    },
  ],
  "stakeNOTEStatus": Object {
    "inCoolDown": false,
    "inRedeemWindow": false,
    "redeemWindowBegin": 1717724243,
    "redeemWindowEnd": 1717983443,
  },
  "systemAccountType": "None",
  "vaultLastUpdateTime": Object {
    "0x30fba4a7ec8591f25b4d37fd79943a4bb6e553e2": 1717159619,
    "0x32d82a1c8618c7be7fe85b2f1c44357a871d52d1": 1712778071,
    "0x84e58d8faa4e3b74d55d9fc762230f15d95570b8": 1711660247,
    "0xb1113cf888a019693b254da3d90f841072d85172": 1717159607,
    "0xd6aa58cf21a0edb33375d6c0434b8bb5b589f021": 1715259311,
    "0xf94507f3dece4cc4c73b6cf228912b85eadc9cfb": 1719005903,
  },
}
`;
